pkg/kubelet/kubeletconfig/controller.go

```diff
@@ -0,0 +1,350 @@
+		return
+	}
+
+	// start the ConfigOK condition sync loop
+	go utilpanic.HandlePanic(func() {
+		utillog.Infof("starting ConfigOK condition sync loop")
+		wait.JitterUntil(func() {
+			cc.configOK.Sync(client, nodeName)
+		}, 10*time.Second, 0.2, true, wait.NeverStop)
+	})()
+
+	// only sync to new, remotely provided configurations if dynamic config was enabled
+	if cc.dynamicConfig {
+		cc.informer = newSharedNodeInformer(client, nodeName,
+			cc.onAddNodeEvent, cc.onUpdateNodeEvent, cc.onDeleteNodeEvent)
+		// start the informer loop
+		// Rather than use utilruntime.HandleCrash, which doesn't actually crash in the Kubelet,
+		// we use HandlePanic to manually call the panic handlers and then crash.
+		// We have a better chance of recovering normal operation if we just restart the Kubelet in the event
+		// of a Go runtime error.
+		go utilpanic.HandlePanic(func() {
+			utillog.Infof("starting Node informer sync loop")
+			cc.informer.Run(wait.NeverStop)
+		})()
+
+		// start the config source sync loop
+		go utilpanic.HandlePanic(func() {
+			utillog.Infof("starting config source sync loop")
+			wait.JitterUntil(func() {
+				cc.syncConfigSource(client, nodeName)
+			}, 10*time.Second, 0.2, true, wait.NeverStop)
+		})()
+	} else {
+		utillog.Infof("dynamic config not enabled, will not sync to remote config")
+	}
+}
+
+// initialize makes sure that the storage layers for various controller components are set up correctly
+func (cc *Controller) initialize() error {
+	utillog.Infof("ensuring filesystem is set up correctly")
+	// initialize local checkpoint storage location
+	if err := cc.checkpointStore.Initialize(); err != nil {
+		return err
+	}
+	// initialize bad config tracker
+	if err := cc.badConfigTracker.Initialize(); err != nil {
+		return err
+	}
+	// initialize startup tracker
+	if err := cc.startupTracker.Initialize(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// localConfig returns the initConfig if it is loaded, otherwise returns the defaultConfig
+func (cc *Controller) localConfig() *componentconfig.KubeletConfiguration {
+	if cc.initConfig != nil {
+		cc.configOK.Set(status.CurInitMessage, status.CurInitOKReason, apiv1.ConditionTrue)
+		return cc.initConfig
+	}
+	cc.configOK.Set(status.CurDefaultMessage, status.CurDefaultOKReason, apiv1.ConditionTrue)
+	return cc.defaultConfig
+}
+
+// inTrial returns true if the time elapsed since the last modification of the current config exceeds `trialDur`, false otherwise
+func (cc *Controller) inTrial(trialDur time.Duration) (bool, error) {
+	now := time.Now()
+	t, err := cc.checkpointStore.CurrentModified()
+	if err != nil {
+		return false, err
+	}
+	if now.Sub(t) > trialDur {
+		return true, nil
+	}
+	return false, nil
+}
+
+// crashLooping returns true if the number of startups since the last modification of the current config exceeds `threshold`, false otherwise
+func (cc *Controller) crashLooping(threshold int32) (bool, error) {
```

Comment:
For (2), if the Kubelet is restarted due to "poor health" within the config trial period, isn't it possible (or even likely) that the "poor health" was caused by a bad configuration?

For (3), that's a good point, kernel panics need to be taken into account. We should probably consider Kubelet restarts since max(last node boot time, last config change time) instead of just the last config change time.